### ISSUE TITLE
Missing symbols: `dlopen`, `dlclose`, `posix_fallocate`

### ISSUE THREAD

**Issue by rice7th** [sunfishcode/eyra#44]
Hi sunfish!
After all the fixes I proposed to get eyra and mustang to work, my tiny miniquad project is still incapable of compiling because `dlopen`, `dlclose` and `posix_fallocate` are all missing. Unfortunately after careful reading I've noticed that dynamic linking is still completely unsupported, so the chance of getting any project involving miniquad or similar crates that bind to system libraries is very minimal. That said, `posix_fallocate` doesn't seem related to dynamic linking at all, so at least that could be a starting point.

If I get time, I might look onto the dynamic linking implementation and see what can I do, but until then my most useful contribution will be this issue.

### Notes

`posix_fallocate` is defined in [`fcntl.h`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_fallocate.html)

`dlopen`, `dlclose` and all the other [`dlfcn.h`](https://pubs.opengroup.org/onlinepubs/7908799/xsh/dlfcn.h.html) functions are already defined in rust inside [`relibc`](https://github.com/redox-os/relibc/blob/master/src/header/dlfcn/mod.rs).

Unfrotunately, I haven't studied the project enough, and something tells me that dynamic linking is quite a bit more challenging to implement than just those functions (I would really love to know what's missing to actually make dynamic linking work), but at least we have a publicly available pure rust implementation as a starting point.

**Comment by sunfishcode** [comment]
Yes; `posix_fallocate` ought to be a fairly straightfoward wrapper around [`rustix::fs::fallocate`](https://docs.rs/rustix/latest/rustix/fs/fn.fallocate.html).

C-style dynamic linking is indeed missing. It's not something I need for my own current projects, so I haven't mapped out what an implementation would look like. @bjorn3 recently [added startup support](https://github.com/sunfishcode/origin/pull/108), but full C-style dynamic linking support would also involve ELF file parsing and loading, relocations processing, dynamic thread-local storage, and possibly more.

**Comment by bjorn3** [comment]
I've been chipping away at getting musl's dynamic linker work with eyra. I've got a working prototype of it already. (see the dynlink branches on my forks or eyra, origin and c-ward. be aware that I force push to them frequently.) There are still several issues with it though. Most of the changes I have done/need to do are things that are almost certainly necessary for a pure rust dynamic linker tok.

**Comment by rice7th** [comment]
@bjorn3 wow that's actually pretty nice! You're doing a lot of useful work; hopefully I might get involved too.

@sunfishcode As for elf parsing and storing, there are actually a **lot** of libraries dedicated to just that. [`goblin`](https://github.com/m4b/goblin) and [`object`](https://github.com/gimli-rs/object). Regarding `goblin`, m4b (the author) also made a [semi functioning dynamic linker](https://github.com/m4b/dyrad) in pure rust.

**Comment by polarathene** [comment]
It might be of interest to others subscribed here, there is a [`dlopen-rs` crate](https://github.com/weizhiao/rust-dlopen) that is pure rust. They're also apparently working on their own `ld.so` equivalent.

**Comment by pothos** [comment]
Yes, thanks for sharing, the author also has https://github.com/weizhiao/rust-elfloader
